axon-server-connector/src/main/java/org/axonframework/axonserver/connector/heartbeat/HeartbeatConfiguration.java

```diff
@@ -0,0 +1,81 @@
+import java.util.function.Function;
+
+import static io.axoniq.axonserver.grpc.control.PlatformOutboundInstruction.RequestCase.HEARTBEAT;
+
+/**
+ * Module configuration that defines the components needed to enable heartbeat and monitor the
+ * availability of the connection with AxonServer.
+ *
+ * @author Sara Pellegrini
+ * @since 4.2
+ */
+public class HeartbeatConfiguration implements ModuleConfiguration {
+
+    private final Function<Configuration, AxonServerConnectionManager> connectionManagerSupplier;
+
+    private final Function<Configuration, AxonServerConfiguration> axonServerConfigurationSupplier;
+
+    private final AtomicReference<Scheduler> heartbeatScheduler = new AtomicReference<>();
+
+    public HeartbeatConfiguration() {
+        this(c -> c.getComponent(AxonServerConnectionManager.class),
+             c -> c.getComponent(AxonServerConfiguration.class));
+    }
+
+    public HeartbeatConfiguration(
+            Function<Configuration, AxonServerConnectionManager> connectionManagerSupplier,
+            Function<Configuration, AxonServerConfiguration> axonServerConfigurationSupplier) {
+        this.connectionManagerSupplier = connectionManagerSupplier;
+        this.axonServerConfigurationSupplier = axonServerConfigurationSupplier;
+    }
+
+    /**
+     * Initializes the {@link GrpcHeartbeatSource} component, needed to send heartbeat to AxonServer,
+     * any time the client will receive an heartbeat from the server.
+     * <p>
+     * Initializes the {@link HeartbeatMonitor} component, needed to force a disconnection if the
+     * communication between the client and the server is no longer available.
+     * <p>
+     * Initializes the {@link Scheduler} used to start and stop the monitor of the connection state.
+     *
+     * @param config the global configuration, providing access to generic components
+     */
+    @Override
+    public void initialize(Configuration config) {
+        AxonServerConnectionManager connectionManager = connectionManagerSupplier.apply(config);
+        AxonServerConfiguration configuration = axonServerConfigurationSupplier.apply(config);
+        String context = configuration.getContext();
+
+        GrpcHeartbeatSource heartbeatSource = new GrpcHeartbeatSource(connectionManager, context);
+        connectionManager.onOutboundInstruction(context, HEARTBEAT, i -> heartbeatSource.send());
+
+        HeartbeatMonitor heartbeatMonitor = new HeartbeatMonitor(connectionManager, context);
+        this.heartbeatScheduler.set(new Scheduler(heartbeatMonitor::run));
+    }
+
+    /**
+     * Starts the monitoring of the connection state.
+     */
+    @Override
+    public void start() {
+        Optional.ofNullable(heartbeatScheduler.get()).ifPresent(Scheduler::start);
+    }
+
+    /**
+     * Stops the monitoring of the connection state.
+     */
+    @Override
+    public void shutdown() {
+        Optional.ofNullable(heartbeatScheduler.get()).ifPresent(Scheduler::stop);
```

Comment:
How big is the chance that the `heartbeatScheduler` `AtomicReference` returns `null`? Should we cover for that by logging something or throwing an exception?

Reply:
The AtomicReference returns null only when the module is not initialized. This should be covered by throwing an exception. I've not found a dedicated exception for this kind of problem, for now I used a RuntimeException. Do you think is better to create a dedicate exception?